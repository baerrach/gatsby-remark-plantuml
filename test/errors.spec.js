const testRemarkPlugin = require(`./test-remark-plugin`)
const { PlantUmlError } = require(`../index`)
const path = require(`path`)
const EOL = require(`os`).EOL

describe(`Error Handling`, () => {
  beforeEach(() => {
    jest.resetModules()
  })

  it(`fails the build when can't find Java and Graphviz`, async () => {
    const path = process.env.PATH
    process.env.PATH = `` // an empty PATH won't find any files
    const code = `
\`\`\`plantuml
@startuml
Alice -> Bob: Authentication Request
Bob --> Alice: Authentication Response

Alice -> Bob: Another authentication Request
Alice <-- Bob: Another authentication Response
@enduml
\`\`\`
    `
    try {
      await testRemarkPlugin.testPlugin({
        code,
        reporter: {
          warn: [
            `Executable 'java' not found on path`,
            `Executable 'dot' not found on path`,
          ],
        },
      })
    } finally {
      process.env.PATH = path
    }
  })

  it(`fails the build when can't find plantuml`, async () => {
    const plantumljar = `./lib/plantuml-does-not-exist.jar`
    const jarFullPath = path.join(process.cwd(), plantumljar)
    const code = `
\`\`\`plantuml
@startuml
Alice -> Bob: Authentication Request
Bob --> Alice: Authentication Response

Alice -> Bob: Another authentication Request
Alice <-- Bob: Another authentication Response
@enduml
\`\`\`
    `
    await testRemarkPlugin.testPlugin({
      code,
      reporter: {
        error: [
          [
            `Could not generate plantuml diagram:  `,
            new Error(`Error: Unable to access jarfile ${jarFullPath}${EOL}`),
          ],
        ],
      },
      options: {
        plantumljar,
      },
    })
  })

  describe(`reports line numbers`, () => {
    it(`on syntax errors`, async () => {
      const code = `
\`\`\`plantuml
@startuml
error here // first line is #1 after @startuml
@enduml
\`\`\`
`

      await testRemarkPlugin.testPlugin({
        code,
        reporter: {
          error: [
            [
              `Could not generate plantuml diagram: Syntax Error at line 1  `,
              new PlantUmlError(`Syntax Error`),
            ],
          ],
        },
      })
    })

    it(`on missing closing quotes`, async () => {
      const code = `
\`\`\`plantuml
@startuml
Alice -> Bob
Bob --> "missing closing quote
@enduml
\`\`\`
`

      await testRemarkPlugin.testPlugin({
        code,
        reporter: {
          error: [
            [
              `Could not generate plantuml diagram: Syntax Error at line 2  `,
              new PlantUmlError(`Syntax Error`),
            ],
          ],
        },
      })
    })
  })
})
