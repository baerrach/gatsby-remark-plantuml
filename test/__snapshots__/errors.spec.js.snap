// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Handling fails the build when can't find Java and Graphviz 1`] = `
Object {
  "children": Array [
    Object {
      "lang": "plantuml",
      "meta": null,
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 10,
          "offset": 203,
        },
        "indent": Array [
          1,
          1,
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "code",
      "value": "@startuml
Alice -> Bob: Authentication Request
Bob --> Alice: Authentication Response

Alice -> Bob: Another authentication Request
Alice <-- Bob: Another authentication Response
@enduml",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 5,
      "line": 11,
      "offset": 208,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Error Handling fails the build when can't find plantuml 1`] = `
Object {
  "children": Array [
    Object {
      "lang": "plantuml",
      "meta": null,
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 10,
          "offset": 203,
        },
        "indent": Array [
          1,
          1,
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "code",
      "value": "@startuml
Alice -> Bob: Authentication Request
Bob --> Alice: Authentication Response

Alice -> Bob: Another authentication Request
Alice <-- Bob: Another authentication Response
@enduml",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 5,
      "line": 11,
      "offset": 208,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Error Handling reports line numbers on missing closing quotes 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "lang": undefined,
      "meta": null,
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 7,
          "offset": 78,
        },
        "indent": Array [
          1,
          1,
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "html",
      "value": "<p><b><span style=\\"color:red\\">PlantUmlError</span></b></p>
<pre>
    0 @startuml
    1 Alice -> Bob
    2 Bob --> \\"missing closing quote
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Syntax Error at line 2 
    3 @enduml
</pre>
",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 8,
      "offset": 79,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Error Handling reports line numbers on missing closing quotes 2`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "lang": undefined,
      "meta": null,
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 7,
          "offset": 78,
        },
        "indent": Array [
          1,
          1,
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "html",
      "value": "<p><b><span style=\\"color:red\\">PlantUmlError</span></b></p>
<pre>
    0 @startuml
    1 Alice -> Bob
    2 Bob --> \\"missing closing quote
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Syntax Error at line 2 
    3 @enduml
</pre>
",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 8,
      "offset": 79,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Error Handling reports line numbers on syntax errors 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "lang": undefined,
      "meta": null,
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 6,
          "offset": 81,
        },
        "indent": Array [
          1,
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "html",
      "value": "<p><b><span style=\\"color:red\\">PlantUmlError</span></b></p>
<pre>
    0 @startuml
    1 error here // first line is #1 after @startuml
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Syntax Error at line 1 
    2 @enduml
</pre>
",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
